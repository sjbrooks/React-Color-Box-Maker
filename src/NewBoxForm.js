import React, { useState } from "react";
// import Box from "./Box";

/** Form for creating a new box in BoxList. Recieves addBox from parent BoxList.
 *
 * Has state for the backgroundColor, width, and height of the box; on submission,
 * sends {backgroundColor, width, height} to fn rec'd from parent.
 *
 */

const NewBoxForm = ({ addBox }) => {
  const INITIAL_STATE = { backgroundColor: "blue", width: 100, height: 100};
  const [formData, setFormData] = useState(INITIAL_STATE);

  /** Send {backgroundColor, width, height} to parent
   *    & clear form. */

  const handleSubmit = evt => {
    evt.preventDefault();
    addBox(formData);
    setFormData(INITIAL_STATE);
  };

  /** Update local state w/curr state of input elem */

  const handleChange = evt => {
    const { backgroundColor, width, height } = evt.target;
    setFormData(fData => ({
      ...fData,

      // NOTE: each of these keys originally looked like arrays
      backgroundColor: backgroundColor,
      width: width,
      height: height
    }));
  };

  /** render form */

  console.log(`\n\n\n The value of formData is `, formData, '\n\n\n');
  return (
    <form onSubmit={handleSubmit}>
      <label htmlFor="backgroundColor">Color:</label>
      <input
        id="backgroundColor"
        name="backgroundColor"
        value={formData.backgroundColor}
        onChange={handleChange}
      />

      <label htmlFor="width">Width:</label>
      <input
        type="number"
        id="width"
        name="width"
        value={formData.width}
        onChange={handleChange}
      />
      
      <label htmlFor="height">Height:</label>
      <input
        type="number"
        id="height"
        name="height"
        value={formData.height}
        onChange={handleChange}
      />

      <button>Add a new box!</button>
    </form>
  );
};

export default NewBoxForm;
