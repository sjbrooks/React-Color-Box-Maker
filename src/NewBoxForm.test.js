import React from "react";
import '@testing-library/jest-dom/extend-expect';
import { render, fireEvent } from "@testing-library/react";
import NewBoxForm from "./NewBoxForm"


test('renders NewBoxForm without crashing', () => {
    render(< NewBoxForm />)
});


test("snapshot:  renders same form each time", function(){
    const{asFragment} = render(<NewBoxForm/>);
    expect(asFragment()).toMatchSnapshot();
})

test('handleChange: can update values for all form fields after form input is changed', function(){
    const{ getByLabelText } = render(< NewBoxForm />);
    const backGroundColorInput = getByLabelText("Color:");
    const widthInput = getByLabelText("Width:");
    const heightInput = getByLabelText("Height:");

    fireEvent.change(backGroundColorInput, { target: {value: "purple"}});
    fireEvent.change(widthInput, { target: {value: "400"}});
    fireEvent.change(heightInput, { target: {value: "12"}});
    
    // QUESTION: we have type:"number" in input tag, but received STRING
    expect(backGroundColorInput.value).toBe("purple");
    expect(widthInput.value).toBe("400");
    expect(heightInput.value).toBe("12");
})

// const testFormData = { backgroundColor: "blue", width: 100, height: 100};
// test("handleSubmit: adds box to page that matches submssion and resets form", function(){
//     const{ container, getByLabelText } = render(< NewBoxForm />);

//     const backGroundColorInput = getByLabelText("Color:");
//     const widthInput = getByLabelText("Width:");
//     const heightInput = getByLabelText("Height:");

//     fireEvent.change(backGroundColorInput, { target: {value: "purple"}});
//     fireEvent.change(widthInput, { target: {value: "400"}});
//     fireEvent.change(heightInput, { target: {value: "12"}});
//     fireEvent.submit(container.querySelector('form'));
    
//     const newBox = container.querySelector('.colored-box');
//     console.log(`\n\n\n The value of newBox is `, newBox);

//     expect(newBox).toContainHTML(testFormData);
// })