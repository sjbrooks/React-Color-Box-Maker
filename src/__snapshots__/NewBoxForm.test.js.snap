// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot:  renders same form each time 1`] = `
<DocumentFragment>
  <form>
    <label
      for="backgroundColor"
    >
      Color:
    </label>
    <input
      id="backgroundColor"
      name="backgroundColor"
      value=""
    />
    <label
      for="width"
    >
      Width:
    </label>
    <input
      id="width"
      name="width"
      type="number"
      value="0"
    />
    <label
      for="height"
    >
      Height:
    </label>
    <input
      id="height"
      name="height"
      type="number"
      value="0"
    />
    <button>
      Add a new box!
    </button>
  </form>
</DocumentFragment>
`;
